contributing/writing-tests.md

```diff
@@ -0,0 +1,199 @@
+expression. Prior to running the tests, provider configuration details such as
+access keys must be made available as environment variables:
+
+```sh
+export HCP_CLIENT_ID=...
+export HCP_CLIENT_SECRET=...
+```
+
+Tests can then be run by specifying a regular expression defining the tests to run:
+
+```sh
+$ make testacc TESTARGS='-run=TestAccConsulCluster'
+TF_ACC=1 go test ./... -v -run=TestAccConsulCluster -timeout 120m
+ok  	github.com/hashicorp/terraform-provider-hcp/internal/consul	(cached) [no tests to run]
+=== RUN   TestAccConsulCluster
+--- PASS: TestAccConsulCluster (538.48s)
+PASS
+ok  	github.com/hashicorp/terraform-provider-hcp/internal/provider	539.112s
+```
+
+Entire resource test suites can be targeted by using the naming convention to
+write the regular expression.
+
+For advanced developers, the acceptance testing framework accepts some additional environment variables that can be used to control Terraform CLI binary selection, logging, and other behaviors. See the [Extending Terraform documentation](https://www.terraform.io/docs/extend/testing/acceptance-tests/index.html#environment-variables) for more information.
+
+## Writing an Acceptance Test
+
+Terraform has a framework for writing acceptance tests which minimises the
+amount of boilerplate code necessary to use common testing patterns. The entry
+point to the framework is the `resource.Test()` function.
+
+Tests are divided into `TestStep`s. Each `TestStep` proceeds by applying some
+Terraform configuration using the provider under test, and then verifying that
+results are as expected by making assertions using the provider API. It is
+common for a single test function to exercise both the creation of and updates
+to a single resource. Most tests follow a similar structure.
+
+1. Pre-flight checks are made to ensure that sufficient provider configuration
+   is available to be able to proceed - in the case of HCP, `HCP_CLIENT_ID` and `HCP_CLIENT_SECRET` must be set prior
+   to running acceptance tests. This is common to all tests exercising a single
+   provider.
+
+Each `TestStep` is defined in the call to `resource.Test()`. Most assertion
+functions are defined out of band with the tests. This keeps the tests
+readable, and allows reuse of assertion functions across different tests of the
+same type of resource. The definition of a complete test looks like this:
+
+```go
+func TestAccConsulCluster(t *testing.T) {
+    resourceName := "hcp_consul_cluster.test"
+
+    resource.Test(t, resource.TestCase{
+        PreCheck:          func() { testAccPreCheck(t) },
+        ProviderFactories: providerFactories,
+        CheckDestroy:      testAccCheckConsulClusterDestroy,
+        Steps: []resource.TestStep{
+            {
+                Config: testConfig(testAccConsulClusterConfig),
+                Check: resource.ComposeTestCheckFunc(
+                    testAccCheckConsulClusterExists(resourceName),
```

Comment:
So this assertion you wrote, right? And the one below is a generic one the provider test suite gives you?

Reply:
Yep, that's correct